src/shared/generator.c

```diff
@@ -833,6 +833,10 @@ int generator_write_cryptsetup_unit_section(
                 "Wants=blockdev@dev-mapper-%%i.target\n"
                 "IgnoreOnIsolate=true\n");
 
+        /* Ensure confexts can be used to customize unlocking disks */
+        if (in_initrd() && !dir_is_empty("/.extra/confext/", /* ignore_hidden_or_backup= */ true))
```

Comment:
Why dir_is_empty() ? Why not to unconditionally add the dependency?

Reply:
If it is dangerous, then why not to make systemd-confext safer??
Not familiar with confext, but does confext fail when there is no confext image to be loaded??
If not, it is not necessary to check here. If so, then confext should be changed to gracefully handle the case.

Reply:
The issue is with imposing ordering constraints that were not there before. There's plenty of conditional checks in generators, it's quite normal and obvious to add such things

Reply:
Actually we have many conditions to add dependencies and so on. But, all (at least most) of the conditions (at least should) come from the configuration source of the unit, e.g. fstab, crypttab and so on.

However, in this case, the condition is comes from the external source. I do not think it is a right way to do.

Reply:
It's the same thing, external configuration resulting in conditional changes. Whether these are expressed via a config file, or a config file contained in an image, is irrelevant, that's just the transport mechanism. It would be extremely wrong to add ordering constraints when they are completely useless and have no effect.

Reply:
So, this is for cryptsetup, but my argument is quite general, thus I will explain by using fstab as an example.

For each generated unit has `SourcePath=/etc/fstab`. And that should be only the source, and the generator should not use any external source files (or add the external ones to `SourcePath=`, but that's ugly). That may not be documented, but I believe the basic rule for the behavior of generators we provided. At least I believe we should do that.
For example, if an fstab entry has `_netdev` option, we do not check if there is a network service that provides connectivity to the network file system, nor check if there exists suitable network configuration files. But, simply add dependency to network-online.target and remote-fs.target. That's why we introduce target units to abstract layers of configuration.

But, in this case, the dependency comes from the external source, i.e. (existence of) image files under the specific directory, and the generated unit does not have any comment about why the dependency is introduced.
I am not familiar with neither confext nor cryptsetup, but this patch deviated from the basic assumption of generated units.

So, even if this will go, please do not silently add the dependency, but add a short comment to the generated unit why the dependency is added, that is, mention that the specific directory is not empty and it may contain a confext image, and so on.

Reply:
> I am not familiar with neither confext nor cryptsetup, but this patch deviated from the basic assumption of generated units.

This is absolutely not how it works, or it ever worked. Just look at gpt-auto, it generates stuff that changes based on the partition GUIDs. The cryptsetup generator changes depending on the paths used by the devices.
Some content also depends on whether or not a UKI was used (eg: TPM dependencies). This is not really any different: whether a UKI is booted or a confext is used depends on the state of the ESP at the time of booting.

You may be thinking about network stuff, but this is completely unrelated. This is how generating stuff for unlocking/setting up the root disks in the initrd always worked, and they _have_ to, as the behaviour changes depending on what's on the inspected disk, on the ESP, on the kernel command line, on credentials from SMBIOS or qemu-fwcg, etc. It is absolutely not the case that only a single configuration file changes these behaviours, and it should most definitely not change now, or it will make it just impossible to use confext/systext to customize the boot process when using UKIs, which is the entire purpose of supporting loading these images from the ESP.

Reply:
Please do not close. I am not convinced yet.

Reply:
We added confext support to the initrd to customize its behaviour. The main purpose of the initrd is to unlock/prep the rootfs. Why do you want to keep it broken?